File: src/main/java/no/novari/flyt/integration/model/dtos/IntegrationDto.java
package no.novari.flyt.integration.model.dtos;

import jakarta.validation.GroupSequence;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;
import no.novari.flyt.integration.model.entities.Integration;
import no.novari.flyt.integration.validation.constraints.ActiveConfigurationIsDefinedIfStateIsActive;
import no.novari.flyt.integration.validation.constraints.ReferencedConfigurationExists;
import no.novari.flyt.integration.validation.constraints.ReferencedConfigurationIsComplete;
import no.novari.flyt.integration.validation.constraints.ReferencedConfigurationIsForIntegration;
import no.novari.flyt.integration.validation.groups.ActiveConfigurationIsCompleteGroup;
import no.novari.flyt.integration.validation.groups.ActiveConfigurationIsForIntegrationGroup;

@GroupSequence({IntegrationDto.class, ActiveConfigurationIsForIntegrationGroup.class, ActiveConfigurationIsCompleteGroup.class})
@Getter
@EqualsAndHashCode
@Jacksonized
@Builder(toBuilder = true)
public class IntegrationDto {

    private long id;

    private Long sourceApplicationId;

    private String sourceApplicationIntegrationId;

    private String destination;

    @ActiveConfigurationIsDefinedIfStateIsActive
    private Integration.State state;

    @ReferencedConfigurationExists
    @ReferencedConfigurationIsForIntegration(groups = ActiveConfigurationIsForIntegrationGroup.class)
    @ReferencedConfigurationIsComplete(groups = ActiveConfigurationIsCompleteGroup.class)
    private Long activeConfigurationId;

}

File: src/main/kotlin/no/novari/flyt/integration/api/dto/IntegrationDto.kt
package no.novari.flyt.integration.api.dto

import com.fasterxml.jackson.annotation.JsonProperty
import jakarta.validation.GroupSequence
import no.novari.flyt.audit.actor.Actor
import no.novari.flyt.integration.persistence.entity.Integration
import no.novari.flyt.integration.validation.constraints.ActiveConfigurationIsDefinedIfStateIsActive
import no.novari.flyt.integration.validation.constraints.ReferencedConfigurationExists
import no.novari.flyt.integration.validation.constraints.ReferencedConfigurationIsComplete
import no.novari.flyt.integration.validation.constraints.ReferencedConfigurationIsForIntegration
import no.novari.flyt.integration.validation.groups.ActiveConfigurationIsCompleteGroup
import no.novari.flyt.integration.validation.groups.ActiveConfigurationIsForIntegrationGroup
import java.time.Instant

@GroupSequence(
    IntegrationDto::class,
    ActiveConfigurationIsForIntegrationGroup::class,
    ActiveConfigurationIsCompleteGroup::class,
)
data class IntegrationDto(
    val id: Long,
    val sourceApplicationId: Long,
    val sourceApplicationIntegrationId: String,
    val destination: String,
    @field:ActiveConfigurationIsDefinedIfStateIsActive
    val state: Integration.State,
    @field:ReferencedConfigurationExists
    @field:ReferencedConfigurationIsForIntegration(groups = [ActiveConfigurationIsForIntegrationGroup::class])
    @field:ReferencedConfigurationIsComplete(groups = [ActiveConfigurationIsCompleteGroup::class])
    val activeConfigurationId: Long? = null,
    @field:JsonProperty(access = JsonProperty.Access.READ_ONLY)
    val createdAt: Instant? = null,
    @field:JsonProperty(access = JsonProperty.Access.READ_ONLY)
    val createdBy: String? = null,
    @field:JsonProperty(access = JsonProperty.Access.READ_ONLY)
    val createdByActor: Actor? = null,
    @field:JsonProperty(access = JsonProperty.Access.READ_ONLY)
    val lastModifiedAt: Instant? = null,
    @field:JsonProperty(access = JsonProperty.Access.READ_ONLY)
    val lastModifiedBy: String? = null,
    @field:JsonProperty(access = JsonProperty.Access.READ_ONLY)
    val lastModifiedByActor: Actor? = null,
)
